hal/src/nRF52840/ble_hal.cpp

```diff
@@ -2426,39 +2469,17 @@ void BleObject::GattServer::processGattServerEvents(const ble_evt_t* event, void
             break;
         }
         case BLE_GATTS_EVT_WRITE: {
-            const ble_gatts_evt_write_t& write = event->evt.gatts_evt.params.write;
-            LOG_DEBUG(TRACE, "BLE GATT Server event: write characteristic, handle: %d", write.handle);
-            BleCharacteristic* characteristic = gatts->findCharacteristic(write.handle);
-            if (characteristic == nullptr) {
-                return;
-            }
-            if (characteristic->charHandles.cccd_handle == write.handle) {
-                if (event->evt.gatts_evt.params.write.data[0] > 0) {
-                    gatts->addToCccdList(characteristic, event->evt.gatts_evt.conn_handle);
-                } else {
-                    gatts->removeFromCCCDList(characteristic, event->evt.gatts_evt.conn_handle);
-                }
-            }
-            // TODO: deal with different write commands.
-            uint8_t* data = (uint8_t*)gatts->gattsDataRecPool_.alloc(write.len);
-            if (!data) {
+            LOG_DEBUG(TRACE, "BLE GATT Server event: data written.");
+            ble_evt_t* dataWrittenEvent = (ble_evt_t*)BleObject::getInstance().dispatcher()->allocEventData(sizeof(ble_evt_t) + \
+                    event->evt.gatts_evt.params.write.len * sizeof(uint8_t));
```

Comment:
This should be `len - 1`, as Nordic SDK uses placeholder arrays of size `[1]`, instead of `[]` or gcc extension `[0]`. There are other cases like this:
https://github.com/particle-iot/device-os/pull/1847/files#diff-fba7ee4853a5cf25cf4e9837d8c5f803R3045
https://github.com/particle-iot/device-os/pull/1847/files#diff-fba7ee4853a5cf25cf4e9837d8c5f803R3067
https://github.com/particle-iot/device-os/pull/1847/files#diff-fba7ee4853a5cf25cf4e9837d8c5f803R3089
https://github.com/particle-iot/device-os/pull/1847/files#diff-fba7ee4853a5cf25cf4e9837d8c5f803R3107
https://github.com/particle-iot/device-os/pull/1847/files#diff-fba7ee4853a5cf25cf4e9837d8c5f803R3141

Reply:
I don't quite understand why this doesn't work, as `event->evt.gatts_evt.params.write.len` should be the total length of the transmitted data.

Reply:
If you go into the definition of `ble_gatts_evt_write_t` you'll see that the `data` field is a variable length array, however it's defined as `data[1]`, so when you allocated `ble_evt_t` you already allocated memory for that one element. The total number of additional bytes that needs to be allocated becomes `len - 1`. This is the case for all the other lines I've mentioned.

Reply:
Nordic SDK even has a macro `SUB1` for these specifically.

Reply:
I got you. In that case, I would still like to allocate the `evt.gatts_evt.params.write.len` of memory to store the data, although there will waste 1-byte length of memory temporary, in case that Nordic references the data using a pointer instead of a placeholder in the future.

Reply:
I'd disagree. They will not be able to easily change this, since they've chosen to use this pattern, and while a waste of 1 byte doesn't seem that much (also arguable since we are using a pool), for all the other cases, it's `sizeof(some_struct_t)` instead.

Reply:
That makes sense to me.